Tech Debt: Fix pre-existing unit test failures blocking CI/CD pipeline

## Problem
The CI/CD pipeline is failing because of multiple pre-existing unit test failures that are blocking all new development. These failures are not related to new features but are technical debt that needs to be addressed.

## Impact
- **Blocking Development**: New features cannot be merged due to failing tests
- **CI/CD Pipeline Failure**: All PRs fail regardless of code quality
- **Developer Productivity**: Time wasted debugging unrelated test failures

## Pre-existing Test Failures Identified

### Module Import Issues
- `src/components/plotpoint/editor/arcane_backgrounds/Editor.test.jsx`
  - **Error**: Cannot find module '../../formgroups/AttributeSelectFormGroup'
  - **Fix**: Update import path or create missing component

### Mock Function Issues  
- `src/components/plotpoint/editor/edges/Editor.test.jsx`
  - **Error**: expect(jest.fn()).toHaveBeenCalledWith(...expected) - Number of calls: 0
  - **Fix**: Review mock setup and event triggering

### Array Length Assertion Issues
- `src/components/character_sheet/selected_skills/index.test.jsx`
  - **Error**: Expected length: 7, Received length: 17
  - **Fix**: Update test expectations or fix component logic

### Async/Timeout Issues
- `src/contexts/AuthContext.test.js`
  - **Error**: Exceeded timeout of 5000 ms, async state update issues
  - **Fix**: Improve async test handling, use proper act() wrapping

### Constructor Issues
- Multiple Editor components
  - **Error**: Class constructor Editor cannot be invoked without 'new'
  - **Fix**: Update test instantiation patterns

### Jest Mock Issues
- `src/App.comprehensive.test.js`
  - **Error**: Invalid variable access: React in jest.mock()
  - **Fix**: Update mock implementation to avoid out-of-scope variables

## Success Criteria
- [ ] All unit tests pass in CI/CD pipeline
- [ ] New PRs can be merged without test interference
- [ ] Test suite runs reliably in under 30 seconds
- [ ] No more `Class constructor cannot be invoked without 'new'` errors
- [ ] All async tests properly wrapped with act()
- [ ] All import paths resolved correctly

## Proposed Solution
1. **Phase 1**: Fix critical blocking tests (imports, constructors)
2. **Phase 2**: Fix async and timeout issues
3. **Phase 3**: Update assertion expectations to match current behavior
4. **Phase 4**: Improve test reliability and speed

## Priority
**High** - Blocking all development workflow

## Acceptance Criteria
- CI/CD pipeline shows green status
- All existing functionality still works
- New features can be developed without test interference

--- a/features/test_suite_pipeline_health.feature
+++ b/features/test_suite_pipeline_health.feature
@@ -1,0 +1,45 @@
+Feature: CI/CD Pipeline Health and Test Suite Reliability
+  As a development team
+  I want all unit tests to pass consistently in the CI/CD pipeline
+  So that new features can be developed and deployed without interference from technical debt
+
+  Background:
+    Given the application has a comprehensive test suite
+    And the CI/CD pipeline runs automated quality checks
+    And developers need reliable feedback on code changes
+
+  Scenario: Unit test suite executes successfully without blocking issues
+    Given the development team has written unit tests for components
+    When the CI/CD pipeline runs the complete test suite
+    Then all unit tests should pass
+    And no import path errors should occur
+    And no mock function configuration issues should prevent execution
+    And no constructor invocation errors should block test runs
+
+  Scenario: Test assertions accurately reflect component behavior
+    Given components have specific expected behaviors
+    When unit tests validate component functionality
+    Then test assertions should match actual component output
+    And array length expectations should align with component logic
+    And component state should be validated correctly
+
+  Scenario: Async test operations complete within reasonable timeframes
+    Given components perform asynchronous operations
+    When tests validate async behavior
+    Then async operations should complete within expected timeout periods
+    And proper test waiting strategies should be implemented
+    And authentication context changes should be handled appropriately
+
+  Scenario: Development workflow remains unblocked by test infrastructure
+    Given developers are implementing new features
+    When they run the test suite locally or in CI/CD
+    Then the test suite should complete in under 30 seconds
+    And all quality gates should pass
+    And new pull requests should not be blocked by pre-existing test failures
+
+  Scenario: Test suite provides reliable development feedback
+    Given the test suite validates business logic and component behavior
+    When developers make changes to the codebase
+    Then test results should accurately reflect the impact of changes
+    And false positive failures should not block development workflow
+    And test execution should be consistent across different environments

--- a/ui-web/features/step_definitions/test_pipeline_steps.js
+++ b/ui-web/features/step_definitions/test_pipeline_steps.js
@@ -1,0 +1,194 @@
+import { Given, When, Then } from 'cucumber';
+import { execSync } from 'child_process';
+
+// Background steps
+Given('the application has a comprehensive test suite', function () {
+  // Verify test files exist
+  const testFiles = execSync('find ui-web/src -name "*.test.js" -o -name "*.test.jsx" | wc -l', { encoding: 'utf-8' });
+  if (parseInt(testFiles.trim()) < 50) {
+    throw new Error('Insufficient test coverage - need comprehensive test suite');
+  }
+});
+
+Given('the CI/CD pipeline runs automated quality checks', function () {
+  // Verify CI/CD configuration exists
+  const hasPackageJsonTests = execSync('grep -q "test:" ui-web/package.json && echo "true" || echo "false"', { encoding: 'utf-8' });
+  if (hasPackageJsonTests.trim() !== 'true') {
+    throw new Error('CI/CD test configuration missing in package.json');
+  }
+});
+
+Given('developers need reliable feedback on code changes', function () {
+  // Business requirement - no technical validation needed
+  this.developmentWorkflowRequirement = 'reliable_feedback';
+});
+
+// Unit test suite execution scenario
+Given('the development team has written unit tests for components', function () {
+  // Verify component tests exist for critical components
+  const componentTests = execSync('find ui-web/src/components -name "*.test.jsx" | wc -l', { encoding: 'utf-8' });
+  if (parseInt(componentTests.trim()) < 20) {
+    throw new Error('Insufficient component test coverage');
+  }
+});
+
+When('the CI/CD pipeline runs the complete test suite', function () {
+  // Execute the test suite (this will reveal current failures)
+  try {
+    this.testResults = execSync('cd ui-web && CI=true npm run test:unit 2>&1', { encoding: 'utf-8' });
+    this.testsPassed = true;
+  } catch (error) {
+    this.testResults = error.stdout || error.message;
+    this.testsPassed = false;
+  }
+});
+
+Then('all unit tests should pass', function () {
+  if (!this.testsPassed) {
+    throw new Error(`Unit tests failed. Output: ${this.testResults}`);
+  }
+});
+
+Then('no import path errors should occur', function () {
+  if (this.testResults.includes('Cannot find module') || this.testResults.includes('Module not found')) {
+    throw new Error('Import path errors detected in test output');
+  }
+});
+
+Then('no mock function configuration issues should prevent execution', function () {
+  if (this.testResults.includes('jest.fn()') && this.testResults.includes('Number of calls: 0')) {
+    throw new Error('Mock function configuration issues detected');
+  }
+});
+
+Then('no constructor invocation errors should block test runs', function () {
+  if (this.testResults.includes('Class constructor') && this.testResults.includes('cannot be invoked without')) {
+    throw new Error('Constructor invocation errors detected in test output');
+  }
+});
+
+// Test assertions scenario
+Given('components have specific expected behaviors', function () {
+  this.componentBehaviorRequirement = 'specific_expected_behaviors';
+});
+
+When('unit tests validate component functionality', function () {
+  // This is covered by the test execution in the previous scenario
+  if (!this.testResults) {
+    throw new Error('Test results not available - run test suite first');
+  }
+});
+
+Then('test assertions should match actual component output', function () {
+  if (this.testResults.includes('Expected') && this.testResults.includes('Received') && !this.testsPassed) {
+    throw new Error('Test assertion mismatches detected - assertions do not match component output');
+  }
+});
+
+Then('array length expectations should align with component logic', function () {
+  if (this.testResults.includes('Expected length:') && this.testResults.includes('Received length:')) {
+    throw new Error('Array length assertion mismatches detected');
+  }
+});
+
+Then('component state should be validated correctly', function () {
+  if (this.testResults.includes('toHaveValue') && this.testResults.includes('Expected') && !this.testsPassed) {
+    throw new Error('Component state validation issues detected');
+  }
+});
+
+// Async operations scenario
+Given('components perform asynchronous operations', function () {
+  this.asyncComponentRequirement = 'async_operations';
+});
+
+When('tests validate async behavior', function () {
+  // Check if async tests are present
+  const asyncTests = execSync('grep -r "async\\|await\\|waitFor" ui-web/src --include="*.test.js" --include="*.test.jsx" | wc -l', { encoding: 'utf-8' });
+  if (parseInt(asyncTests.trim()) < 5) {
+    throw new Error('Insufficient async test coverage');
+  }
+});
+
+Then('async operations should complete within expected timeout periods', function () {
+  if (this.testResults.includes('Exceeded timeout') || this.testResults.includes('timeout of')) {
+    throw new Error('Async timeout issues detected in test output');
+  }
+});
+
+Then('proper test waiting strategies should be implemented', function () {
+  if (this.testResults.includes('act()') && this.testResults.includes('Warning')) {
+    throw new Error('Improper async test handling detected - missing act() wrappers');
+  }
+});
+
+Then('authentication context changes should be handled appropriately', function () {
+  if (this.testResults.includes('AuthContext') && !this.testsPassed) {
+    throw new Error('Authentication context test issues detected');
+  }
+});
+
+// Development workflow scenario
+Given('developers are implementing new features', function () {
+  this.developmentContext = 'feature_implementation';
+});
+
+When('they run the test suite locally or in CI/CD', function () {
+  // Record test execution time
+  const startTime = Date.now();
+  try {
+    execSync('cd ui-web && CI=true timeout 45 npm run test:unit', { encoding: 'utf-8' });
+    this.testExecutionTime = Date.now() - startTime;
+    this.testsCompletedInTime = true;
+  } catch (error) {
+    this.testExecutionTime = Date.now() - startTime;
+    this.testsCompletedInTime = false;
+  }
+});
+
+Then('the test suite should complete in under 30 seconds', function () {
+  if (this.testExecutionTime > 30000) {
+    throw new Error(`Test suite took ${this.testExecutionTime}ms - exceeds 30 second requirement`);
+  }
+});
+
+Then('all quality gates should pass', function () {
+  if (!this.testsPassed) {
+    throw new Error('Quality gates failing - test suite not passing');
+  }
+});
+
+Then('new pull requests should not be blocked by pre-existing test failures', function () {
+  if (!this.testsPassed) {
+    throw new Error('Pre-existing test failures are blocking development workflow');
+  }
+});
+
+// Test reliability scenario
+Given('the test suite validates business logic and component behavior', function () {
+  this.businessLogicValidation = 'required';
+});
+
+When('developers make changes to the codebase', function () {
+  this.codeChanges = 'simulated';
+});
+
+Then('test results should accurately reflect the impact of changes', function () {
+  // Verify tests are deterministic and not flaky
+  if (this.testResults.includes('Jest has detected') && this.testResults.includes('open handles')) {
+    throw new Error('Test suite has resource leaks - may cause flaky results');
+  }
+});
+
+Then('false positive failures should not block development workflow', function () {
+  if (!this.testsPassed && this.testResults.includes('Warning')) {
+    throw new Error('Test warnings may be causing false positive failures');
+  }
+});
+
+Then('test execution should be consistent across different environments', function () {
+  // Check for environment-specific test issues
+  if (this.testResults.includes('ENOENT') || this.testResults.includes('permission denied')) {
+    throw new Error('Environment-specific test execution issues detected');
+  }
+});

--- a/ui-web/src/components/plotpoint/editor/BasicRules.jsx
+++ b/ui-web/src/components/plotpoint/editor/BasicRules.jsx
@@ -24,10 +24,10 @@ export default class BasicRules extends React.Component {
 		},
 	}
 
-	maximumAttributePointsChange = e => this.props.onChange(Object.assign({}, this.props.basicRules, {maximumAttributePoints: parseInt(e.target.value, 10)}))
-	maximumMajorHindrancesChange = e => this.props.onChange(Object.assign({}, this.props.basicRules, {maximumMajorHindrances: parseInt(e.target.value, 10)}))
-	maximumMinorHindrancesChange = e => this.props.onChange(Object.assign({}, this.props.basicRules, {maximumMinorHindrances: parseInt(e.target.value, 10)}))
-	maximumSkillPointsChange     = e => this.props.onChange(Object.assign({}, this.props.basicRules, {maximumSkillPoints: parseInt(e.target.value, 10)}))
+	maximumAttributePointsChange = e => this.props.onChange(Object.assign({}, this.props.basicRules, {maximumAttributePoints: parseInt(e.target.value, 10) || 0}))
+	maximumMajorHindrancesChange = e => this.props.onChange(Object.assign({}, this.props.basicRules, {maximumMajorHindrances: parseInt(e.target.value, 10) || 0}))
+	maximumMinorHindrancesChange = e => this.props.onChange(Object.assign({}, this.props.basicRules, {maximumMinorHindrances: parseInt(e.target.value, 10) || 0}))
+	maximumSkillPointsChange     = e => this.props.onChange(Object.assign({}, this.props.basicRules, {maximumSkillPoints: parseInt(e.target.value, 10) || 0}))
 
 	render () {
 		let {id, basicRules} = this.props

--- a/ui-web/src/components/plotpoint/editor/BasicRules.test.jsx
+++ b/ui-web/src/components/plotpoint/editor/BasicRules.test.jsx
@@ -1,5 +1,6 @@
 import React from 'react';
 import { render, screen, fireEvent } from '@testing-library/react';
+import '@testing-library/jest-dom';
 import BasicRules from './BasicRules';
 
 // Mock bootstrap-react-components

--- a/ui-web/src/components/plotpoint/editor/arcane_backgrounds/Editor.jsx
+++ b/ui-web/src/components/plotpoint/editor/arcane_backgrounds/Editor.jsx
@@ -2,7 +2,7 @@ import {NumberFormGroup, TextAreaFormGroup, TextFormGroup} from 'bootstrap-react
 import PropTypes from 'prop-types'
 import React                                               from 'react'
 import BaseEditor                                          from '../../../BaseEditor'
-import AttributeSelectFormGroup                            from '../../formgroups/AttributeSelectFormGroup'
+import AttributeSelectFormGroup                            from '../../../formgroups/AttributeSelectFormGroup'
 
 export default class Editor extends React.Component {
 
@@ -14,8 +14,8 @@ export default class Editor extends React.Component {
 		this.props.onDelete(this.props.index)
 	}
 	skillNameChange           = e => this.props.onChange(Object.assign({}, this.props.item, {skillName: e.target.value}), this.props.index)
-	startingPowersChange      = e => this.props.onChange(Object.assign({}, this.props.item, {startingPowers: parseInt(e.target.value, 10)}), this.props.index)
-	startingPowerPointsChange = e => this.props.onChange(Object.assign({}, this.props.item, {startingPowerPoints: parseInt(e.target.value, 10)}), this.props.index)
+	startingPowersChange      = e => this.props.onChange(Object.assign({}, this.props.item, {startingPowers: parseInt(e.target.value, 10) || 0}), this.props.index)
+	startingPowerPointsChange = e => this.props.onChange(Object.assign({}, this.props.item, {startingPowerPoints: parseInt(e.target.value, 10) || 0}), this.props.index)
 
 	render() {
 		let componentId = `ArcaneBackgroundEditor-${this.props.index}-${this.props.id}`

--- a/ui-web/src/components/plotpoint/editor/armor/Editor.jsx
+++ b/ui-web/src/components/plotpoint/editor/armor/Editor.jsx
@@ -4,7 +4,7 @@ import MundaneItemEditor from '../gear/mundane_items/Editor'
 
 export default class Editor extends MundaneItemEditor {
 
-	onArmorChange = e => this.props.onChange(Object.assign({}, this.props.item, {armor: parseInt(e.target.value, 10)}), this.props.index)
+	onArmorChange = e => this.props.onChange(Object.assign({}, this.props.item, {armor: parseInt(e.target.value, 10) || 0}), this.props.index)
 
 
 	additionalFields = () => <div id={'ArmorEditorComponent-' + this.props.id}>

--- a/ui-web/src/components/plotpoint/editor/beasts/Editor.jsx
+++ b/ui-web/src/components/plotpoint/editor/beasts/Editor.jsx
@@ -19,7 +19,7 @@ export default class Editor extends React.Component {
 		selected: ''
 	}
 
-	armorChange            = e => this.props.onChange(Object.assign({}, this.props.item, {armor: parseInt(e.target.value, 10)}), this.props.index)
+	armorChange            = e => this.props.onChange(Object.assign({}, this.props.item, {armor: parseInt(e.target.value, 10) || 0}), this.props.index)
 	onChange               = item => this.props.onChange(item, this.props.index)
 	specialAbilitiesChange = specialAbilities => {this.props.onChange(Object.assign({}, this.props.item, {specialAbilities}), this.props.index)}
 

--- a/ui-web/src/components/plotpoint/editor/gear/mundane_items/Editor.jsx
+++ b/ui-web/src/components/plotpoint/editor/gear/mundane_items/Editor.jsx
@@ -8,8 +8,8 @@ export default class Editor extends GearEditor {
 		id: 'MundaneItemEditorComponent'
 	}
 
-	costChange   = e => this.props.onChange(Object.assign({}, this.props.item, {cost: parseInt(e.target.value, 10)}), this.props.index)
-	weightChange = e => this.props.onChange(Object.assign({}, this.props.item, {weight: parseInt(e.target.value, 10)}), this.props.index)
+	costChange   = e => this.props.onChange(Object.assign({}, this.props.item, {cost: parseInt(e.target.value, 10) || 0}), this.props.index)
+	weightChange = e => this.props.onChange(Object.assign({}, this.props.item, {weight: parseInt(e.target.value, 10) || 0}), this.props.index)
 
 	additionalFields = () => <div>
 

--- a/ui-web/src/components/plotpoint/editor/powers/Form.jsx
+++ b/ui-web/src/components/plotpoint/editor/powers/Form.jsx
@@ -29,7 +29,7 @@ export default class Form extends React.Component {
 		event.preventDefault();
 		this.props.onDelete(this.props.index);
 	};
-	powerPointChange  = e => this.props.onChange(Object.assign({}, this.props.item, {powerPoints: parseInt(e.target.value, 10)}), this.props.index);
+	powerPointChange  = e => this.props.onChange(Object.assign({}, this.props.item, {powerPoints: parseInt(e.target.value, 10) || 0}), this.props.index);
 	rangeChange       = e => this.props.onChange(Object.assign({}, this.props.item, {range: e.target.value}), this.props.index);
 	rankChange        = e => this.props.onChange(Object.assign({}, this.props.item, {rank: e.target.value}), this.props.index);
 

--- a/ui-web/src/contexts/AuthContext.test.js
+++ b/ui-web/src/contexts/AuthContext.test.js
@@ -844,6 +844,7 @@ describe('AuthContext - Remember Me Functionality', () => {
 
   describe('token validation with remember me', () => {
     it('checks token expiry on context initialization with remember me', async () => {
+      jest.setTimeout(15000);
       api.getAccessToken.mockReturnValue('mock-token');
       tokenUtils.isRememberMeSession.mockReturnValue(true);
       tokenUtils.getStoredTokenExpiry.mockReturnValue(Date.now() + 60000);
@@ -865,23 +866,25 @@ describe('AuthContext - Remember Me Functionality', () => {
         );
       };
 
-      render(
-        <AuthProvider>
-          <TestComponent />
-        </AuthProvider>
-      );
+      await act(async () => {
+        render(
+          <AuthProvider>
+            <TestComponent />
+          </AuthProvider>
+        );
+      });
 
       await waitFor(() => {
         expect(authContext).toBeTruthy();
         expect(authContext.loading).toBe(false);
-      }, { timeout: 3000 });
+      }, { timeout: 10000 });
 
       await waitFor(() => {
         expect(authContext.user).toEqual({
           id: 1,
           email: 'test@example.com'
         });
-      }, { timeout: 3000 });
+      }, { timeout: 10000 });
 
       expect(tokenUtils.isRememberMeSession).toHaveBeenCalled();
       expect(tokenUtils.getStoredTokenExpiry).toHaveBeenCalled();

